File: backend/aiHandler.js
const path = require("path");
const fs = require("fs");
const {
  pipeline,
  AutoTokenizer,
  AutoModelForSeq2SeqLM,
} = require("@xenova/transformers");

const MODEL_DIR = path.resolve(__dirname, "model");
let classifier = null;

const verifyModelFiles = () => {
  const requiredFiles = ["tokenizer.json", "config.json"];
  for (const file of requiredFiles) {
    const filePath = path.join(MODEL_DIR, file);
    if (!fs.existsSync(filePath)) {
      throw new Error(`❌ Missing model file: ${file}`);
    }
  }
};

const loadModel = async () => {
  if (classifier) {
    console.log("✅ AI Model already loaded.");
    return;
  }

  try {
    verifyModelFiles();

    const modelURI = `file:///${MODEL_DIR.replace(/\\/g, "/")}`;
    console.log(`🤖 Loading AI model locally from: ${modelURI}`);

    const tokenizer = await AutoTokenizer.from_pretrained(modelURI, {
      local_files_only: true,
    });

    const model = await AutoModelForSeq2SeqLM.from_pretrained(modelURI, {
      local_files_only: true,
    });

    classifier = await pipeline("text2text-generation", model, tokenizer);
    console.log("✅ AI Model loaded successfully (local)!");
  } catch (error) {
    console.error("❌ Failed to load AI model locally:", error.message);
    console.log("🌐 Attempting to download model remotely...");

    try {
      const tokenizer = await AutoTokenizer.from_pretrained("Xenova/distilbart-cnn-6-6");
      const model = await AutoModelForSeq2SeqLM.from_pretrained("Xenova/distilbart-cnn-6-6");
      classifier = await pipeline("text2text-generation", model, tokenizer);
      console.log("✅ Model downloaded and loaded successfully!");
    } catch (err) {
      console.error("❌ Remote model load failed too:", err.message);
      classifier = null;
    }
  }
};

const getAIResponse = async (userMessage) => {
  if (!classifier) {
    await loadModel();
    if (!classifier) throw new Error("AI model not available.");
  }

  const prompt = `
You are a helpful healthcare assistant AI.
Provide clear, concise, and safe responses.
Always include a disclaimer to consult a doctor.

User: ${userMessage}
Answer:
`;

  try {
    const result = await classifier(prompt, {
      max_new_tokens: 256,
      temperature: 0.5,
    });
    return result[0].generated_text;
  } catch (error) {
    console.error("AI generation error:", error.message);
    throw new Error("AI response error");
  }
};

module.exports = { loadModel, getAIResponse };

File: backend/ai_handler.py
# -*- coding: utf-8 -*-
import os
import sys
import torch
from transformers import AutoTokenizer, AutoModelForSeq2SeqLM

# Optional Windows fix for Unicode console
try:
    sys.stdout.reconfigure(encoding='utf-8')
except Exception:
    pass

# ✅ Correct path for your fine-tuned model
MODEL_DIR = os.path.join(os.path.dirname(__file__), "model", "flan_t5_healthcare", "checkpoint-4000")

model = None
tokenizer = None


def load_model():
    """Load the fine-tuned healthcare model"""
    global model, tokenizer
    if model and tokenizer:
        print("[FLASK] Model already loaded.")
        return model, tokenizer

    try:
        print(f"[FLASK] Loading fine-tuned model from {MODEL_DIR} ...")
        tokenizer = AutoTokenizer.from_pretrained(MODEL_DIR, local_files_only=True)
        model = AutoModelForSeq2SeqLM.from_pretrained(MODEL_DIR, local_files_only=True)
        model.eval()
        print("[FLASK] Model loaded successfully from local files.")
    except Exception as e:
        print(f"[FLASK] Local model load failed: {e}")
        print("[FLASK] Attempting to load fallback model (Flan-T5 base)...")
        tokenizer = AutoTokenizer.from_pretrained("google/flan-t5-base")
        model = AutoModelForSeq2SeqLM.from_pretrained("google/flan-t5-base")
        model.eval()
        print("[FLASK] Fallback model loaded successfully.")

    return model, tokenizer


def generate_response(user_message: str) -> str:
    """Generate a text response from the healthcare AI model"""
    global model, tokenizer
    if not model or not tokenizer:
        load_model()

    # Construct clear healthcare prompt
    prompt = (
        f"You are CareCompanion, a professional AI healthcare assistant.\n"
        f"Provide accurate, concise, and safe advice.\n"
        f"Always include a disclaimer reminding to consult a medical professional.\n\n"
        f"User: {user_message}\nAI:"
    )

    inputs = tokenizer(prompt, return_tensors="pt", truncation=True).to(model.device)

    with torch.no_grad():
        outputs = model.generate(
            **inputs,
            max_new_tokens=256,
            temperature=0.5,
            top_p=0.9,
            do_sample=True,
            repetition_penalty=1.2,
            pad_token_id=tokenizer.eos_token_id,
        )

    reply = tokenizer.decode(outputs[0], skip_special_tokens=True).strip()

    if "consult" not in reply.lower():
        reply += "\n\nDisclaimer: This information is for educational purposes only. Always consult a qualified doctor."

    return reply
